### PR TITLE
Instrument mod_bosh

### PR DESCRIPTION
**Pull request by gustawlippa**
Add instrumentation to `mod_bosh`.
I think there are two non-obvious things in this PR:
1. Why not use `instrument/1` and let the new mongoose_instrument modules deal with the event registration?

	The reason is that when running with multiple host_types, instrumentation callback would be run multiple times, per host type. However, `mod_bosh` acts more like a service, in that its handlers are global. Therefore, the BOSH handlers don't have access to which host types they actually run for, and therfore I decided that the instrumentation will omit the host_type label. But if we omit this label, we have to ensure that event registration is run only once - globally.
2. Why change the type of the BOSH socket supervisor?

    I've noticed that after adding the call to `mongoose_instrument:tear_down/1` in `mod_bosh` the tests started failing. It turned out, that when `mod_bosh` was stopped, only the handlers were stopped, but the whole supervision tree remained in place. As the supervision tree is the thing that `mod_bosh` tests to see if it has already been started once (as described in 1.), it was problematic for tests, which have to restart `mod_bosh` multiple times. They started failing, because, simplifying:
    1. Test group started `mod_bosh` and registered instrumentation events. 
    2. Test group ended and `mod_bosh` was stopped, as well as instrumentation events were deregistered.
    5. Next test group started and tried to start `mod_bosh`. Since it checked that the supervisor was up, it didn't register the instrumentation events.
    6. Test failed, because the events weren't registered and trying to execute them failed.
    
	Maybe the other solution would be to check something else when starting `mod_bosh`, but I think the proposed solution is clean, and follows the expectations when stopping a module.
	
Note: presence_one test in metrics_c2s_SUITE on the mssql_mnesia_26 preset has been really flaky, but has finally passed after a few tries. I don't see a connection to this PR.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[small_tests_25](https://circleci.com/gh/esl/MongooseIM/219089) / small_tests / c68f0c00506f6b03574104d09d9a67ca4d28f137
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4281/219089/small_tests.25.3.2.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4281/219089/small_tests.25.3.2.9-amd64/small/ct_run.mongooseim@localhost.2024-05-20_10.28.26/lib.mongooseim.logs/run.2024-05-20_10.28.27/suite.log.html)

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/219130) / elasticsearch_and_cassandra_mnesia / e1df308ac270ee2bd9fd348682bab6ff1050c4a4
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4281/219130/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4281/219130/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%409469ebba6cd5.2024-05-20_10.31.39/index.html)
**OK: 457** / Failed: 0 / **User-skipped: 41** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/219166) / elasticsearch_and_cassandra_mnesia / 960c34e4363629c4f0077f8518f67b3d63db81c1
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4281/219166/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4281/219166/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%409d3eabd0e425.2024-05-20_10.47.27/index.html)
**OK: 457** / Failed: 0 / **User-skipped: 41** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/219214) / elasticsearch_and_cassandra_mnesia / 991d238ea71289fb62b7b0b10e26d453b2678b89
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4281/219214/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4281/219214/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40ebfbb6030cee.2024-05-20_13.42.50/index.html)
**OK: 457** / Failed: 0 / **User-skipped: 41** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/219238) / elasticsearch_and_cassandra_mnesia / e5c2363dab66f3b16e8931f981fcce4ec129ad23
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4281/219238/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4281/219238/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40cee8430fc4c9.2024-05-20_13.43.38/index.html)
**OK: 457** / Failed: 0 / **User-skipped: 41** / Auto-skipped: 0

File: src/mod_bosh.erl [hunk 1]
```diff
@@ -178,9 +190,8 @@ info(forward_body, Req, S) ->
     forward_body(Req1, BodyElem, S#rstate{req_sid = Sid});
 info({bosh_reply, El}, Req, S) ->
     BEl = exml:to_binary(El),
-    %% for BOSH 'data.xmpp.sent.raw' metric includes 'body' wrapping elements
-    %% and resending attempts
-    mongoose_metrics:update(global, [data, xmpp, sent, c2s, bosh], byte_size(BEl)),
+    %% 'mod_bosh_data_sent' metric includes 'body' wrapping elements and resending attempts
+    mongoose_instrument:execute(mod_bosh_data_sent, #{}, #{byte_size =>  byte_size(BEl)}),
```
Comment (chrzaszcz) [inline src/mod_bosh.erl:194]:
```suggestion
    mongoose_instrument:execute(mod_bosh_data_sent, #{}, #{byte_size => byte_size(BEl)}),
```

File: src/mod_bosh.erl [hunk 2]
```diff
@@ -304,8 +315,8 @@ forward_body(Req, #xmlel{} = Body, #rstate{opts = Opts} = S) ->
 
 -spec handle_request(pid(), event_type(), exml:element()) -> ok.
 handle_request(Socket, EventType, Body) ->
-    %% for BOSH 'data.xmpp.received.raw' metric includes 'body' wrapping elements
-    mongoose_metrics:update(global, [data, xmpp, received, c2s, bosh], exml:xml_size(Body)),
+    %% 'mod_bosh_data_received' metric includes 'body' wrapping elements
+    mongoose_instrument:execute(mod_bosh_data_received, #{}, #{byte_size =>  exml:xml_size(Body)}),
```
Comment (chrzaszcz) [inline src/mod_bosh.erl:319]:
```suggestion
    mongoose_instrument:execute(mod_bosh_data_received, #{}, #{byte_size => exml:xml_size(Body)}),
```

**Review by chrzaszcz — COMMENTED** [review]
Looks good!

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/219363) / elasticsearch_and_cassandra_mnesia / e53b65d61f9732647c47b02e997582588b3d62f4
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4281/219363/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4281/219363/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40bde8a61bdfc6.2024-05-21_09.35.31/index.html)
**OK: 457** / Failed: 0 / **User-skipped: 41** / Auto-skipped: 0

**Review by chrzaszcz — APPROVED** [review]
Looks good 👍

**Pull request merged**